### PR TITLE
refactor: improve message serialization

### LINKED ISSUE #1030: fix message serialization in Mellea
We are utilizing our Message class both as a regular component and as a way to serialize our components into chat messages inference engines understand.

We need to
1. Fix our backend serialization pipeline so that we aren't printing Messages twice due to document handling
2. Add infrastructure for better message serialization (either from components directly or utilizing an intermediate message representation). This will allow use to write custom serialization when required (ie for tool results, etc...).

Related is allowing components to specify roles: https://github.com/generative-computing/mellea/issues/377. Presumably anything else that can be serialized as a chat message to an inference provider might be worth allowing components to expose (ie images?).

### PR DESCRIPTION
**Pull request by AngeloDanducci**
# Pull Request

## Issue
Fixes #1030

## Description

Parses MOTs into Messages, honor component declared metadata across all backends

### Testing
- [x] Tests added to the respective file if code was changed
- [ ] New code has 100% coverage if code was added
- [ ] Ensure existing tests and github automation passes (a maintainer will kick off the github automation when the rest of the PR is populated)

### Attribution
- [x] AI coding assistants used

---

### Adding a new component, requirement, sampling strategy, or tool?

If your PR adds or modifies one of the types below, check the matching box. A checklist of type-specific review items will be posted as a comment.

- [ ] Component
- [ ] Requirement
- [ ] Sampling Strategy
- [ ] Tool

NOTE: Please ensure you have an issue that has been acknowledged by a core contributor and routed you to open a pull request against this repository. Otherwise, please open an issue before continuing with this pull request.

### CODE REVIEW

**Comment by AngeloDanducci** [comment]
While this covers all of the explicit portions of the issue, I have created a follow up with what I see as a way forward to be a bit more flexible in the future.

Design discussion here: https://github.com/generative-computing/mellea/issues/1565 cc @jakelorocco

File: mellea/core/base.py [hunk 1]
```diff
@@ -1242,6 +1226,47 @@ async def astream(self) -> str:
...
+                # /`Message` action path uses) so its role, tool calls, and reasoning
+                # survive rather than degrading to the raw string value.
+                # Deferred import: `chat` imports `mellea.core`, so a top-level import
+                # here would be circular. `Message._parse` reads only its `computed`
+                # argument, so a throwaway `Message` instance is a safe vehicle.
+                from ..stdlib.components.chat import Message
+
+                self.parsed_repr = Message(role="assistant", content=self.value)._parse(
+                    self
+                )  # type: ignore[assignment]
```
Comment (jakelorocco) [inline mellea/core/base.py:1350]: Ignore this comment. I am going to just repost it to the design discussion. I think it's beyond the scope of this PR and isn't necessary to improve the current behavior like this PR does.

~This case may require additional handling (and it may be enough to note it in the design discussion).~

If I do:
```
class C1(Component[C2]):
  ...

class C2(Component)[C3]:
  ...
```

Then generating from C1, should create a MoT[C2]; ie a ModelOutputThunk that parses as a C2 component.

I don't believe our current generation code does this, but that would potentially mean:
```
m1 = act(C1, ...)
assert isinstance(m1.parsed_repr, C2)

m2 = act(m1, ...)
assert isinstance(m2.parsed_repr, C3)
```

I think the important aspect of this is that the mot should probably still get output exactly as it was returned by the model, but have different generation implications.

File: test/backends/test_tool_serialization_unit.py [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+
+
+def _tool_ctx(tool_msg: Message) -> ChatContext:
+    """A context with an assistant tool-call turn followed by `tool_msg`."""
+    return (
+        ChatContext()
+        .add(Message("user", "use a tool"))
+        .add(Message("assistant", "calling tool", tool_calls=_TOOL_CALLS))
+        .add(tool_msg)
+    )
```
Comment (jakelorocco) [inline test/backends/test_tool_serialization_unit.py:48]: How do providers handle messages where multiple tools are requested at once? If this was a mot here (instead of the tool call assistant message), would it get serialized a single message or multiple tool call messages?

Comment (AngeloDanducci) [reply test/backends/test_tool_serialization_unit.py:48]: My current understanding:

A single mot holds all the tools requested in that turn (https://github.com/generative-computing/mellea/blob/main/mellea/core/base.py#L805). 

When that mot is parsed into a Message (https://github.com/generative-computing/mellea/blob/main/mellea/stdlib/components/chat.py#L211)):

- ollama: build_tool_calls(computed) loops the whole list and returns one list[ToolCallDict] (https://github.com/generative-computing/mellea/blob/main/mellea/helpers/openai_compatible_helpers.py#L436), assigned to a single Message.tool_calls.
- openai/watsonx/litellm: we take the provider's own msg["tool_calls"] array verbatim (https://github.com/generative-computing/mellea/blob/main/mellea/stdlib/components/chat.py#L223), which is already multi-entry.
- HF: falls back to the raw string (https://github.com/generative-computing/mellea/blob/main/mellea/stdlib/components/chat.py#L226)

A two-tool turn produces tool_calls=[{...call_1}, {...call_2}] on one assistant turn, and each result comes back as its own role="tool" turn keyed by tool_call_id (Watsonx/HF) or tool_name (Ollama). 

One message with tool_calls being an N-entry array.

File: mellea/backends/ollama.py [hunk 3]
```diff
@@ -68,6 +69,37 @@ def _strip_data_uri_prefix(images: list[str]) -> list[str]:
     return stripped
 
 
+def _to_ollama_tool_calls(openai_tool_calls: list[dict[str, Any]]) -> list[dict]:
```
Comment (markstur) [inline mellea/backends/ollama.py:72]: Recommendation
Add unit tests for _to_ollama_tool_calls() in test/backends/test_ollama_unit.py to cover:
- Valid JSON string arguments (already covered indirectly)
- Invalid JSON → empty dict (critical for robustness)
- Dict arguments passthrough
- Edge cases (empty strings, missing keys)
The existing integration-style tests in test_tool_serialization_unit.py verify the end-to-end behavior, but direct unit tests would catch regressions in the translation logic itself.

File: mellea/backends/ollama.py [hunk 4]
```diff
@@ -68,6 +69,37 @@ def _strip_data_uri_prefix(images: list[str]) -> list[str]:
...
+    Returns:
+        The same tool calls in Ollama's native shape.
+    """
+    translated: list[dict] = []
+    for tc in openai_tool_calls:
+        fn = tc.get("function", {}) or {}
+        raw_args = fn.get("arguments", {})
+        if isinstance(raw_args, str):
+            try:
+                args = json.loads(raw_args) if raw_args else {}
```
Comment (markstur) [inline mellea/backends/ollama.py]: ???  So here and below it will silently turn args that fail to parse or fail type check into {}.

It would be safer to do error handling for that.  Otherwise a misread of tool args, could be turned into a tool call w/ default args and that behavior should be worse than an intentional error.

If we want to go with "it's probably safe enough" then I'd say at least add logging and docstring.

**Review by markstur — CHANGES_REQUESTED** [review]
I think silently eating args that cannot be parsed (or wrong type) should be addressed.

I'd vote for error, but if in practice it is better as-is (e.g. unlikely), then probably at least logging.

File: mellea/backends/ollama.py (hunk 4 above)
Comment (AngeloDanducci) [reply mellea/backends/ollama.py]: Originally I was thinking arguments should always be valid JSON otherwise they'd end up as a JSONDecodeError but that's probably too opimistic.

Changed to updating the docstring and logging - similar to what `_extract_model_tool_requests` does.

File: mellea/backends/ollama.py (hunk 3 above)
Comment (AngeloDanducci) [reply mellea/backends/ollama.py:72]: Added additional coverage that should cover these.

**Review by markstur — APPROVED** [review]
LGTM

My comments addressed.

Jake's comments look addressed or "self-ignored".

**Review by jakelorocco — APPROVED** [review]
